src/smtencoding/FStar.SMTEncoding.Encode.fs

```diff
@@ -1054,15 +1054,60 @@ and encode_sigelt' (env:env_t) (se:sigelt) : (decls_t * env_t) =
         | _ -> false) in
        decls@rest@inversions, env
 
-     | Sig_inductive_typ(t, _, tps, k, _, datas) ->
+     | Sig_inductive_typ(t, universe_names, tps, k, _, datas) ->
+         let tcenv = env.tcenv in
+         let is_injective  =
+             let usubst, uvs = SS.univ_var_opening universe_names in
+             let env, tps, k =
+                Env.push_univ_vars tcenv uvs,
+                SS.subst_binders usubst tps,
+                SS.subst (SS.shift_subst (List.length tps) usubst) k
+             in
+             let tps, k = SS.open_term tps k in
+             let _, k = U.arrow_formals k in //don't care about indices here
+             let tps, env_tps, _, us = TcTerm.tc_binders env tps in
+             let u_k =
+                match (SS.compress k).n with
+                | Tm_type u -> u
+                | Tm_fvar fv when S.fv_eq_lid fv FStar.Parser.Const.eqtype_lid -> U_zero
+                | _ -> failwith (BU.format1 "Impossible: Type of inductive is %s" (Print.term_to_string k))
+             in
+             //BU.print2 "Universe of tycon: %s : %s\n" (Ident.string_of_lid t) (Print.univ_to_string u_k);
+             let rec universe_leq u v =
```

Comment:
I didn't want to put this in Syntax.Util since I didn't want it to be used outside of a context (e.g., in the typechecker) where universe unification variables may still be around. This boolean function may give the wrong result when called on universe terms with U_unif terms. The proper way of checking of enforcing universe inequalities in the typechecker would be to call FStar.TypeChecker.Rel to register and solve universe inequality constraints.

Reply:
Note, within the SMT encoding, we really ought to not have U_unif anymore ... I suppose I could assert that here.

Reply:
OTOH, when this function returns false, the is_injective flag is false ... i.e., erring conservatively.

Reply:
I indeed interpreted as a conservative leq, where returning false means "I don't know". It might be useful somewhere else, but we could factor it out when the need arises.